### DISCUSSION TITLE
Public Versions vs Published Objects

### DISCUSSION THREAD

**Discussion by cancan101** [pimcore/pimcore#16740] [Q&A / Support]
What is the difference between a `public` version and a published object? More specifically will the public version be the most recent version of the object at which time it was published? I am trying to understand this code: https://github.com/pimcore/pimcore/blob/4da5b84117965399f87d6151b765f25cddb01a6c/lib/Maintenance/Tasks/VersionsCleanupTask.php#L127-L132

and specifically if it will keep around the published version of the object.

**Comment by dpfaffenbauer** [comment]
Yes, so basically, when you click "Save & Publish", you basically create a new version and publish that version. If you only click "Save a new Version", Pimcore just creates a new version and does not publish it.

**Reply by cancan101** [reply]
Where / how does the "Save & Publish" mark that version as public? Assuming the `/save` endpoint gets called with `task=publish`, I don't see where the new version with public is set here: https://github.com/pimcore/admin-ui-classic-bundle/blob/cd5cb36beda37d87059d3eb7f88a8b393c98492a/src/Controller/Admin/DataObject/DataObjectController.php#L1289-L1374

In addition it looks like the `changePublishedState: "force_published"` on workflows does not create a new (public) version, which seems inconsistent:
https://github.com/pimcore/pimcore/blob/4da5b84117965399f87d6151b765f25cddb01a6c/lib/Workflow/EventSubscriber/ChangePublishedStateSubscriber.php#L54-L56

**Comment by dpfaffenbauer** [comment]
Pimcore internally does that somehow, not sure where exactly. 

Workflow: I am pretty certain that there might be inconsistency

**Comment by fashxp** [comment]
As far as I know, public versions are only available in combination with documents: 
![image](https://github.com/pimcore/pimcore/assets/8792145/a85604b4-7f52-4cdf-9c91-f3bc54645ed5)

and can be used for A/B testing

**Reply by cancan101** [reply]
It looks like there is some concept of which version was published, even if it isn't the "public" field:
![image](https://github.com/pimcore/pimcore/assets/51059/e327b5d0-1b72-4a34-ae86-eb100064f684)
how is the specific published version tracked?

it looks like some combination of this: https://github.com/pimcore/admin-ui-classic-bundle/blob/cd5cb36beda37d87059d3eb7f88a8b393c98492a/src/Controller/Admin/DataObject/DataObjectController.php#L396-L399
and:
https://github.com/pimcore/admin-ui-classic-bundle/blob/cd5cb36beda37d87059d3eb7f88a8b393c98492a/public/js/pimcore/asset/versions.js#L89-L91

though I don't totally understand how this logic works and what it relies on being set in the DB. Ultimately I am trying to understand: How do I prevent the version that would be marked public from getting deleted by the cleanup task?
